pinot-core/src/main/java/org/apache/pinot/core/periodictask/BasePeriodicTask.java

```diff
@@ -29,21 +33,31 @@
 @ThreadSafe
 public abstract class BasePeriodicTask implements PeriodicTask {
   private static final Logger LOGGER = LoggerFactory.getLogger(BasePeriodicTask.class);
+  private static final String DEFAULT_REQUEST_ID = "auto";
 
   // Wait for at most 30 seconds while calling stop() for task to terminate
   private static final long MAX_PERIODIC_TASK_STOP_TIME_MILLIS = 30_000L;
 
   protected final String _taskName;
   protected final long _intervalInSeconds;
   protected final long _initialDelayInSeconds;
+  protected final ReentrantLock _runLock;
 
   private volatile boolean _started;
   private volatile boolean _running;
 
+  // Properties that task may use during execution. null by default.
+  protected Properties _activePeriodicTaskProperties;
```

Comment:
(Major)
Keep swapping the `_activePeriodicTaskProperties` and expect the child class to use it to identify the table name is hard to manage.
We should add an abstract method `runTask(Properties periodicTaskProperties)` and the child class can implement accordingly

Reply:
Amrish  had previously done the approach of `runTask(Properties)` . It did not seem best since each sub-task had to be modified. 
I think things have changed enough that we can take one more look at that. @amrishlal maybe you can have a quick discussion with Jackie and point out the problems. I am fine with either approach.

Reply:
We can make it call `runTask()` by default, and override it for `ControllerPeriodicTask`. In `ControllerPeriodicTask`, both of them should call `processTables()` internally

Reply:
Pending...

Reply:
Since this is a multi-threaded class, we want to have all the threads that want to do the same thing (i.e run a task) pass through the same execution chain using the same state variables. Doing this will avoid a lot of confusion with respect to what thread is executing where and which state variables that thread is using and how and where locks are being acquired and released. The alternative (which I was trying earlier as Subbu mentioned) seemed to be leading to a situation where we would not only have multiple threads, but also multiple paths of execution (even if that path is just a few functions deep and converges later on) and multiple references to the same state variable (base class has _activePeriodicTaskProperties, do we pass that on the stack as well to child classes when it is already available to all child classes. If not, then why are we passing the "other" properties object on stack, etc). Basically, it was bringing up too many issues and ending up with making too many assumptions; hence, I abandoned the approach.

The main problem here is that we need to replace the active _activePeriodicTaskProperties object and that is really all that needs to be done to make the same code work for all threads who want to do the same thing, so the simplest way to do that seemed to be was to create a run(properties) method to swap out _activePeriodicTaskProperties under lock and then call run() method to let the rest of the code do what it is already doing.

Reply:
Logically they are 2 different tasks: one periodic trigger to process all tables lead by the controller; the other one adhoc to process based on the custom properties. The second one might only need to process one table, why mixing them and force the second one to scan potentially thousands of tables?
Also, I don't agree that for multi-threaded class, we want to follow the same execution chain and mixing the logics. It won't really help with the race condition handling, and more complicated logic within a single method can lead to more bugs.